lib/charms/saml_integrator/v0/saml.py

```diff
@@ -0,0 +1,253 @@
+        url_key = ""
+        for key in relation_data:
+            if key.endswith("_redirect_url") or key.endswith("_post_url"):
+                url_key = key
+        lowercase_name = "_".join(url_key.split("_")[:-2])
+        name = "".join(x.capitalize() for x in lowercase_name.split("_"))
+        http_method = url_key.split("_")[-2]
+        prefix = f"{lowercase_name}_{http_method}_"
+        return cls(
+            name=name,
+            url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}url"]),
+            binding=relation_data[f"{prefix}binding"],
+            response_url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}response_url"])
+            if f"{prefix}response_url" in relation_data
+            else None,
+        )
+
+
+class SamlRelationData(BaseModel):
+    """Represent the relation data.
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: List of SAML certificates.
+        endpoints: List of SAML endpoints.
+    """
+
+    entity_id: str = Field(..., min_length=1)
+    metadata_url: AnyHttpUrl
+    certificates: List[str]
+    endpoints: List[SamlEndpoint]
+
+    def to_relation_data(self) -> Dict[str, str]:
+        """Convert an instance of SamlDataAvailableEvent to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        result = {
+            "entity_id": self.entity_id,
+            "metadata_url": self.metadata_url,
+            "x509certs": ",".join(self.certificates),
+        }
+        for endpoint in self.endpoints:
+            result = {**result, **endpoint.to_relation_data()}
+        return result
+
+
+class SamlDataAvailableEvent(ops.RelationEvent):
+    """Saml event emitted when relation data has changed.
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: List of SAML certificates.
+        endpoints: List of SAML endpoints.
+    """
+
+    @property
+    def entity_id(self) -> str:
+        """Fetch the SAML entity ID from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("entity_id")
+
+    @property
+    def metadata_url(self) -> str:
+        """Fetch the SAML metadata URL from the relation."""
+        assert self.relation.app
+        return parse_obj_as(AnyHttpUrl, self.relation.data[self.relation.app].get("metadata_url"))
+
+    @property
+    def certificates(self) -> List[str]:
+        """Fetch the SAML certificates from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("x509certs").split(",")
+
+    @property
+    def endpoints(self) -> List[SamlEndpoint]:
+        """Fetch the SAML endpoints from the relation."""
+        endpoints = []
+        assert self.relation.app
+        relation_data = self.relation.data[self.relation.app]
+        for key in relation_data:
```

Comment:
Can this be converted to a list comprehension?

Reply:
Done